### ISSUE TITLE
RFC: Add end-to-end vision-language model support to SDPOTrainer

### ISSUE THREAD

**Issue by DimensionSTP** [huggingface/trl#6832]
### Feature request

`SDPOTrainer` accepts a `ProcessorMixin` and can load a vision-language model, but multimodal inputs are not preserved end to end.

With a tiny Qwen3.5 VLM fixture, the processor correctly creates:

- `pixel_values`
- `image_grid_thw`
- `mm_token_type_ids`

The current SDPO flow then:

- returns only `input_ids` from `_tokenize_prompts_untruncated`;
- passes `images=None` to `VLLMGeneration.generate`;
- reconstructs successful-rollout teacher prompts without the original image block;
- calls student and teacher forwards without multimodal kwargs;
- uses generic buffered tensor splitting even though packed Qwen pixel rows need grid-aware split/unsplit handling.

I propose adding end-to-end VLM support to SDPO with a deliberately bounded first scope:

- same-architecture VLM student and local teacher;
- one source image per conversational prompt;
- image-aware Transformers and vLLM generation;
- successful-rollout reprompting that preserves the original structured image block;
- independent processor outputs for the original student context and reprompted teacher context;
- image tensors forwarded to both student and teacher;
- distillation loss aligned only on completion tokens;
- grid-aware split/unsplit for packed vision tensors across buffered micro-batches;
- unchanged text-only behavior.

The first change should fail fast for unsupported VLM teacher-server scoring rather than silently dropping images. Cross-architecture teachers, tool-returned images, and arbitrary multi-image prompts can remain follow-ups.

### Motivation

The current API can initialize with a VLM processor/model, but the image disappears at multiple later stages. Fixing only `images=None` would be unsafe: rollout generation could become image-conditioned while policy, student, or teacher log-probability computations remain text-only. Likewise, preserving the image in reprompt messages without correctly splitting packed pixel tensors can misalign image tokens and features.

A CPU-only reproducer against main at `06c89977492db9102c41a920b499426de6b1c776` produced:

```text
processor_image_keys=['image_grid_thw', 'mm_token_type_ids', 'pixel_values']
sdpo_tokenize_return_type=list
vllm_images=None
student_forward_keys=['attention_mask', 'input_ids', 'logits_to_keep', 'use_cache']
reprompt_image_block_count=0
result=REPRODUCED
```

The tiny fixture itself completes a real CPU VLM forward with finite logits, so the result is not caused by a broken processor or model fixture. The affected source paths remain unchanged on current main `3f738b90ed9de71c76f0593f575ad62851c30ba1`.

Related work:

- #6650 added VLM plumbing to `DistillationTrainer` and provides useful processor and packed-tensor patterns.
- #6515 adds VLM support to `AsyncGRPOTrainer`.
- Neither change handles SDPO's successful-rollout reprompting and student/local-teacher self-distillation flow.

Questions for maintainers:

1. Is same-architecture, single-image, local-teacher SDPO the right first support boundary?
2. Should the first PR support both Transformers and colocate vLLM generation, or land the Transformers path first?
3. Is an explicit fail-fast for VLM teacher-server mode preferable until that request protocol carries images?

### Your contribution

I can implement the agreed first scope and add focused tests for processor field preservation, image-aware generation, image-preserving reprompting, both model forwards, completion alignment, packed pixel splitting, and text-only regression.

I have prepared a standalone CPU reproducer and validated the tiny Qwen3.5 VLM fixture with a real forward pass. AI assistance was used during investigation and drafting; I reviewed the reproduction, source analysis, and proposed scope directly.

**Comment by DimensionSTP** [comment]
@qgallouedec, could you advise on the preferred first scope for this proposal?

Since this RFC was opened, Async GRPO VLM support landed in #6839, alongside the existing DistillationTrainer VLM support in #6650. SDPO still drops image inputs in its generation and successful-rollout reprompting paths, so those changes do not resolve this proposal.

Would you prefer an initial PR limited to single-image, same-architecture student/local-teacher training with Transformers generation, leaving vLLM generation for a follow-up? This would still preserve images through reprompting, both forwards, and buffered batching, while keeping text-only behavior unchanged. Thanks!
